feat: add private properties to policy definition entity

suh-rao opened the pull request:
## What this PR changes/adds

Add private properties to policy definition entity. The implementation supports in-memory and postgres as a store.

## Why it does that

The private properties are required to provide the owner with the possibility to add extra properties which should not be exposed outside. This requirement was initiated when we want to add administrative properties like (createdBy, createdAt, chnagedBy, chnagedAt), policyType (Access, usage). We assume that the private properties can be used to store such a requirement after registering a listener on the policy definition.
More information is in this discussion:
https://github.com/eclipse-edc/Connector/discussions/2592#discussioncomment-7379257

## Further notes

_List other areas of code that have changed but are not necessarily linked to the main feature. This could be method
signature changes, package declarations, bugs that were encountered and were fixed inline, etc._

## Linked Issue(s)


Closes https://github.com/eclipse-edc/Connector/issues/3567

jimmarino commented:
@suh-rao Can you please not open-and-close PRs like this as it puts an administrative burden on the reviewers? Stick to one PR per topic.

suh-rao commented:
> @suh-rao Can you please not open-and-close PRs like this as it puts an administrative burden on the reviewers? Stick to one PR per topic.

@jimmarino : ECA was not updated with old PR even after retrigger. Hence need to close and reopen it again. Will stick to this PR now for the feature

paullatzelsperger commented:
Removing my review as I am out of office

jimmarino commented:
Add @paullatzelsperger to review when he is back since this is an area he has worked on extensively.

paullatzelsperger reviewed (changes requested):
still missing:
- a test where a Policy is queried by a private property
- E2E tests in the `PolicyDefinitionApiEndToEndTest`. It's not needed to re-test all test cases,  tests for creating policies with private props, querying with private props, one update scenario and a delete scenario would suffice.

suh-rao commented:
> still missing:
> 
> * a test where a Policy is queried by a private property
> * E2E tests in the `PolicyDefinitionApiEndToEndTest`. It's not needed to re-test all test cases,  tests for creating policies with private props, querying with private props, one update scenario and a delete scenario would suffice.

>>>a test where a Policy is queried by a private property
I followed approach done for contractdefinition. In this PR can I only do changes regarding storing privateProperties as JSON and query part later in next PR?

suh-rao commented:
> > still missing:
> > 
> > * a test where a Policy is queried by a private property
> > * E2E tests in the `PolicyDefinitionApiEndToEndTest`. It's not needed to re-test all test cases,  tests for creating policies with private props, querying with private props, one update scenario and a delete scenario would suffice.
> 
> > > > a test where a Policy is queried by a private property
> 
> I followed approach done for contractdefinition. In this PR can I only do changes regarding storing privateProperties as JSON and query part later in next PR?

Hello, I am unable to add label to pill request.
Also is it ok to have query part later? (As mentioned in above point)

ndr-brt reviewed (approved):
I'm ok with tackling the query in a subsequent pr, please open an issue related to it

suh-rao commented on extensions/control-plane/api/management-api/policy-definition-api/src/test/java/org/eclipse/edc/connector/api/management/policy/transform/JsonObjectFromPolicyDefinitionTransformerTest.java:
```diff
@@ -60,4 +69,21 @@ void transform() {
     }
 
+    @Test
+    void transform_withPrivateProperties_simpleTypes() {
```
I added relevant to test store complexType too.

suh-rao commented:
> I'm ok with tackling the query in a subsequent pr, please open an issue related to it

https://github.com/eclipse-edc/Connector/issues/3623. This is new issue created for search

suh-rao commented:
@paullatzelsperger

paullatzelsperger commented:
> > I'm ok with tackling the query in a subsequent pr, please open an issue related to it
> 
> #3623. This is new issue created for search

Hey @suh-rao thanks for creating the issue, however it's not an adoption request, just a feature request. Adoption requests are for bigger features, for example "Support XYZ Database in EDC". Please adapt the description of #3623 thus.

paullatzelsperger reviewed (changes requested):
I already said it in my previous review: please extend `PolicyDefinitionApiEndToEndTest`, so that it uses private properties, and add corresponding test assertions.

Also, Checkstyle is failing.

suh-rao commented:
> > > I'm ok with tackling the query in a subsequent pr, please open an issue related to it
> > 
> > 
> > #3623. This is new issue created for search
> 
> Hey @suh-rao thanks for creating the issue, however it's not an adoption request, just a feature request. Adoption requests are for bigger features, for example "Support XYZ Database in EDC". Please adapt the description of #3623 thus.

Changed issue description as "feature request"

suh-rao commented:
> > > I'm ok with tackling the query in a subsequent pr, please open an issue related to it
> > 
> > 
> > #3623. This is new issue created for search
> 
> Hey @suh-rao thanks for creating the issue, however it's not an adoption request, just a feature request. Adoption requests are for bigger features, for example "Support XYZ Database in EDC". Please adapt the description of #3623 thus.

Hey @paullatzelsperger I updated issue description 

> I already said it in my previous review: please extend `PolicyDefinitionApiEndToEndTest`, so that it uses private properties, and add corresponding test assertions.
> 
> Also, Checkstyle is failing.

Hey @paullatzelsperger I updated endtoend tests also checkstyle issues. All workflows should be green now. Please have a look at into it.

paullatzelsperger commented:
> > > > I'm ok with tackling the query in a subsequent pr, please open an issue related to it
> > > 
> > > 
> > > #3623. This is new issue created for search
> > 
> > 
> > Hey @suh-rao thanks for creating the issue, however it's not an adoption request, just a feature request. Adoption requests are for bigger features, for example "Support XYZ Database in EDC". Please adapt the description of #3623 thus.
> 
> Hey @paullatzelsperger I updated issue description
> 
> > I already said it in my previous review: please extend `PolicyDefinitionApiEndToEndTest`, so that it uses private properties, and add corresponding test assertions.
> > Also, Checkstyle is failing.
> 
> Hey @paullatzelsperger I updated endtoend tests also checkstyle issues. All workflows should be green now. Please have a look at into it.

dependency check still failing.

suh-rao commented:
> > > > > I'm ok with tackling the query in a subsequent pr, please open an issue related to it
> > > > 
> > > > 
> > > > #3623. This is new issue created for search
> > > 
> > > 
> > > Hey @suh-rao thanks for creating the issue, however it's not an adoption request, just a feature request. Adoption requests are for bigger features, for example "Support XYZ Database in EDC". Please adapt the description of #3623 thus.
> > 
> > 
> > Hey @paullatzelsperger I updated issue description
> > > I already said it in my previous review: please extend `PolicyDefinitionApiEndToEndTest`, so that it uses private properties, and add corresponding test assertions.
> > > Also, Checkstyle is failing.
> > 
> > 
> > Hey @paullatzelsperger I updated endtoend tests also checkstyle issues. All workflows should be green now. Please have a look at into it.
> 
> dependency check still failing.

In this PR I havent changed any dependencies. So not sure what can be the issue. I synched the branch to main now.

paullatzelsperger commented:
> In this PR I havent changed any dependencies. So not sure what can be the issue. I synched the branch to main now.

@suh-rao just take a look at the failing check, you need to replace the contents of your `DEPENDENCIES` file with whatever the job output says. unfortunately, this can happen without any of your (or our, for that matter) doing.

suh-rao commented:
> > In this PR I havent changed any dependencies. So not sure what can be the issue. I synched the branch to main now.
> 
> @suh-rao just take a look at the failing check, you need to replace the contents of your `DEPENDENCIES` file with whatever the job output says. unfortunately, this can happen without any of your (or our, for that matter) doing.

@paullatzelsperger All 17 checks passed now. Sync to main branch resolved the issue with dependency.

suh-rao commented:
> > > In this PR I havent changed any dependencies. So not sure what can be the issue. I synched the branch to main now.
> > 
> > 
> > @suh-rao just take a look at the failing check, you need to replace the contents of your `DEPENDENCIES` file with whatever the job output says. unfortunately, this can happen without any of your (or our, for that matter) doing.
> 
> @paullatzelsperger All 17 checks passed now. Sync to main branch resolved the issue with dependency.

Hello @paullatzelsperger : can it be merged now? I can work on search by properties feature after that.
